### DISCUSSION TITLE
How do different types of Git reset differ?

### DISCUSSION THREAD

**Discussion by NadChel** [community/community#68224] [Programming Help] [answered]
### Body

I try to understand the difference between `reset --soft`, `reset --mixed`, and `reset --hard`. Docs seem to be incomplete or even misleading. For example, the doc page [says](https://git-scm.com/docs/git-reset) this about `reset --soft`:

> Does not touch the index file or the working tree at all (but resets the head to <commit>, just like all modes do).

But it does affect the staging area (referred to as "the index file"): all the changes between the last commit and the new HEAD pointer go to the staging area. I discovered it by experimenting on my own. My overall takeaways are these:

* `git reset --soft` moves all the changes to the staging area. Versioned files are unchanged
* `git reset --mixed` (default) keeps the versioned files unchanged, but the changes are not automatically moved to the staging area. After executing `git add`, it's identical to `git reset --soft`
* `git reset --hard` destroys all the changes: they are neither in the files (whose content is changed back to what it was at the time of the commit you reset to) nor in the staging area. There's no coming back for your changes

Is my understanding correct?

### Guidelines

- [x] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by seyLu** [comment, marked as answer]
#### Hi @Nadchel,

The way I think about it is:
`git reset --hard`: undos `git commit`, `git add`, and changes you've made since last commit
`git reset --mixed`: undos `git commit` and `git add`
`git reset --soft`: undos `git commit`

**Reply by NadChel** [reply]
Thank you! Is that a yes?

**Reply by seyLu** [reply]
Yup, your understanding is correct. But if I were to nitpick:
> git reset --hard destroys all the changes: they are neither in the files (whose content is changed back to what it was at the time of the commit you reset to) nor in the staging area. There's no coming back for your changes

You can still technically [undo `git reset --hard`](https://stackoverflow.com/questions/5473/how-can-i-undo-git-reset-hard-head1) by resetting to a sha returned by `git reflog`.
